app/src/main/kotlin/com/wire/android/ui/home/settings/backup/BackupAndRestoreViewModel.kt

```diff
@@ -1,17 +1,272 @@
+            delay(300)
+
+            when (val result = createBackupUseCase.invoke(password)) {
+                is CreateBackupResult.Success -> {
+                    state = state.copy(backupCreationProgress = BackupCreationProgress.Finished)
+                    latestCreatedBackup = BackupAndRestoreState.CreatedBackup(
+                        result.backupFilePath,
+                        result.backupFileName,
+                        result.backupFileSize,
+                        password.isNotEmpty()
+                    )
+                }
+                is CreateBackupResult.Failure -> {
+                    state = state.copy(backupCreationProgress = BackupCreationProgress.Failed)
+                    appLogger.e("Failed to create backup: ${result.coreFailure}")
+                }
+            }
+        }
+    }
+
+    fun saveBackup() = viewModelScope.launch {
+        latestCreatedBackup?.let { backupData ->
+            fileManager.shareWithExternalApp(backupData.path, backupData.assetName.fileExtension()) {}
+        }
+        state = BackupAndRestoreState.INITIAL_STATE
+    }
+
+    fun chooseBackupFileToRestore(uri: Uri) = viewModelScope.launch {
+        val importedBackupPath = kaliumFileSystem.tempFilePath(TEMP_IMPORTED_BACKUP_FILE_NAME)
+        uri.copyToTempPath(context, importedBackupPath)
+
+        extractBackupFiles(importedBackupPath)
+
+        // Delete the imported backup file
+        kaliumFileSystem.delete(importedBackupPath)
+    }
+
+    private fun showPasswordDialog() {
+        state = state.copy(restoreFileValidation = RestoreFileValidation.PasswordRequired)
+    }
+
+    @Suppress("MagicNumber")
+    private suspend fun extractBackupFiles(importedBackupPath: Path) {
+        when (val result = extractCompressedFileUseCase.invoke(importedBackupPath)) {
+            is ExtractCompressedBackupFileResult.Success -> {
+                if (result.isEncrypted) {
+                    latestExtractedBackupRootPath = result.extractedFilesRootPath
+                    showPasswordDialog()
+                } else {
+                    importDatabase(result.extractedFilesRootPath)
+                }
+            }
+            is ExtractCompressedBackupFileResult.Failure -> {
+                state = state.copy(restoreFileValidation = RestoreFileValidation.IncompatibleBackup)
+                appLogger.e("Failed to extract backup files: ${result.error}")
+            }
+        }
+    }
+
+    private suspend fun importDatabase(extractedBackupRootPath: Path) {
+        state = state.copy(
+            restoreFileValidation = RestoreFileValidation.ValidNonEncryptedBackup,
+            backupRestoreProgress = BackupRestoreProgress.InProgress(0.75f)
+        )
+        when (importBackup(extractedBackupRootPath, null)) {
+            RestoreBackupResult.Success -> {
+                state = state.copy(backupRestoreProgress = BackupRestoreProgress.InProgress(0.75f))
+                delay(300)
```

Comment:
This delay is not guarantied I think

Reply:
I talked with @gongracr  here and this is just to create an effect of a progress otherwise we do not have a progess, from what I understood this will be added later to this feature